File: utils/git/history.py
from rich.panel import Panel
from rich import box
from typing import Optional

class GitHistory:
    """
    Advanced Git history manipulation: time, author, and message rewriting.
    """
    # This class expects run_command, console, and _get_modification_scope to be available.

    def _get_modification_scope(self, title: str) -> str:
        """
        Prompt the user to select the breadth of a history modification.
        """
        self.console.print(Panel(
            f"[bold cyan]{title}[/]\n\n"
            f"[1] Last Commit Only\n"
            f"[2] Specific Commit (by Hash)\n"
            f"[3] All Commits in Branch", 
            box=box.ROUNDED
        ))
        choice = self.console.input("[bold white]Select scope (1/2/3): [/]")
        if choice == "2": return "specific"
        if choice == "3": return "all"
        return "single"

    def change_commit_time(self, date_str: Optional[str] = None) -> bool:
        """
        Update the timestamp for one or more commits.
        """
        scope = self._get_modification_scope("Time Warp Controller")
        
        target_hash = ""
        if scope == "specific":
            target_hash = self.console.input("[bold white]Enter commit hash: [/]").strip()

        if not date_str:
            date = self.console.input("[bold white]Select commit date (YYYY-MM-DD): [/]")
            time = self.console.input("[bold white]Select commit time (HH:MM, 24h): [/]")
            date_str = f"{date} {time}:00"

        if scope in ["all", "specific"]:
            filter_script = f"export GIT_AUTHOR_DATE='{date_str}'; export GIT_COMMITTER_DATE='{date_str}'"
            if scope == "specific":
                filter_script = f"case \"$GIT_COMMIT\" in {target_hash}*) {filter_script} ;; esac"
            
            cmd = ["filter-branch", "-f", "--env-filter", filter_script, "--", "HEAD"]
            status_msg = "[bold yellow]Rewriting history...[/]"
        else:
            env = {"GIT_AUTHOR_DATE": date_str, "GIT_COMMITTER_DATE": date_str}
            cmd = ["commit", "--amend", "--no-edit", "--date", date_str]
            status_msg = f"[bold yellow]Updating last commit to {date_str}...[/]"

        with self.console.status(status_msg):
            success = self.run_command(cmd, env=None if scope in ["all", "specific"] else env)
        
        if success:
            detail = f"Target: [bold]{target_hash if scope == 'specific' else scope.capitalize()}[/]"
            self.console.print(Panel(
                f"{detail}\nNew Timestamp: [bold yellow]{date_str}[/]",
                title="Time Warp", border_style="yellow", box=box.ROUNDED
            ))
        return success

    def change_commit_author(self) -> bool:
        """
        Interactively update the author identity for one or more commits.
        """
        scope = self._get_modification_scope("Identity Swapper")
        
        target_hash = ""
        if scope == "specific":
            target_hash = self.console.input("[bold white]Enter commit hash: [/]").strip()

        name = self.console.input("[bold white]Enter author name: [/]")
        email = self.console.input("[bold white]Enter author email: [/]")
        author_str = f"{name} <{email}>"

        if scope in ["all", "specific"]:
            filter_script = f"export GIT_AUTHOR_NAME='{name}'; export GIT_AUTHOR_EMAIL='{email}'; export GIT_COMMITTER_NAME='{name}'; export GIT_COMMITTER_EMAIL='{email}'"
            if scope == "specific":
                filter_script = f"case \"$GIT_COMMIT\" in {target_hash}*) {filter_script} ;; esac"
                
            cmd = ["filter-branch", "-f", "--env-filter", filter_script, "--", "HEAD"]
            status_msg = "[bold yellow]Rewriting author in history...[/]"
        else:
            cmd = ["commit", "--amend", "--no-edit", f"--author={author_str}"]
            status_msg = f"[bold yellow]Changing last commit identity...[/]"

        with self.console.status(status_msg):
            success = self.run_command(cmd)
        
        if success:
            detail = f"Target: [bold]{target_hash if scope == 'specific' else scope.capitalize()}[/]"
            self.console.print(Panel(
                f"{detail}\nNew Author: [bold cyan]{author_str}[/]",
                title="Identity Swapped", border_style="cyan", box=box.ROUNDED
            ))
        return success

    def change_commit_message(self) -> bool:
        """
        Interactively update the commit message of the most recent commit.
        """
        self.console.print("[bold cyan]Message Rewriter[/]")
        new_msg = self.console.input("[bold white]Enter new commit message: [/]")

        with self.console.status("[bold green]Updating message...[/]"):
            success = self.run_command(["commit", "--amend", "-m", new_msg])
        
        if success:
            self.console.print(Panel(
                f"New Message: [bold white]{new_msg}[/]",
                title="Message Updated", border_style="green", box=box.ROUNDED
            ))
        return success

File: internal/git/history.go
package git

import (
	"bytes"
	"errors"
	"fmt"
	"os/exec"
	"strings"
)

type GitHistory struct {
	Core *GitCore
}

func NewGitHistory(core *GitCore) (*GitHistory, error) {
	if core == nil {
		return nil, errors.New("git core is required")
	}
	return &GitHistory{Core: core}, nil
}

func (g *GitHistory) ChangeCommitTime(scope, targetHash, dateTime string) error {
	scope, targetHash, err := normalizeScope(scope, targetHash)
	if err != nil {
		return err
	}

	dateTime = strings.TrimSpace(dateTime)
	if dateTime == "" {
		return errors.New("dateTime cannot be empty")
	}

	if scope == "single" {
		env := map[string]string{
			"GIT_AUTHOR_DATE":    dateTime,
			"GIT_COMMITTER_DATE": dateTime,
		}
		if _, stderr, err := g.runCommandWithEnv(env, "commit", "--amend", "--no-edit", "--date", dateTime); err != nil {
			if stderr != "" {
				return fmt.Errorf("failed to amend commit time: %s", stderr)
			}
			return fmt.Errorf("failed to amend commit time: %w", err)
		}
		return nil
	}

	filterScript := buildTimeFilterScript(scope, targetHash, dateTime)
	if _, stderr, err := g.Core.RunCommand("filter-branch", "-f", "--env-filter", filterScript, "--", "HEAD"); err != nil {
		if stderr != "" {
			return fmt.Errorf("failed to rewrite commit time history: %s", stderr)
		}
		return fmt.Errorf("failed to rewrite commit time history: %w", err)
	}

	return nil
}

func (g *GitHistory) ChangeCommitAuthor(scope, targetHash, name, email string) error {
	scope, targetHash, err := normalizeScope(scope, targetHash)
	if err != nil {
		return err
	}

	name = strings.TrimSpace(name)
	email = strings.TrimSpace(email)
	if name == "" {
		return errors.New("author name cannot be empty")
	}
	if email == "" {
		return errors.New("author email cannot be empty")
	}

	if scope == "single" {
		author := fmt.Sprintf("%s <%s>", name, email)
		if _, stderr, err := g.Core.RunCommand("commit", "--amend", "--no-edit", "--author", author); err != nil {
			if stderr != "" {
				return fmt.Errorf("failed to amend commit author: %s", stderr)
			}
			return fmt.Errorf("failed to amend commit author: %w", err)
		}
		return nil
	}

	filterScript := buildAuthorFilterScript(scope, targetHash, name, email)
	if _, stderr, err := g.Core.RunCommand("filter-branch", "-f", "--env-filter", filterScript, "--", "HEAD"); err != nil {
		if stderr != "" {
			return fmt.Errorf("failed to rewrite commit author history: %s", stderr)
		}
		return fmt.Errorf("failed to rewrite commit author history: %w", err)
	}

	return nil
}

func (g *GitHistory) ChangeCommitMessage(newMessage string) error {
	newMessage = strings.TrimSpace(newMessage)
	if newMessage == "" {
		return errors.New("commit message cannot be empty")
	}

	if _, stderr, err := g.Core.RunCommand("commit", "--amend", "-m", newMessage); err != nil {
		if stderr != "" {
			return fmt.Errorf("failed to amend commit message: %s", stderr)
		}
		return fmt.Errorf("failed to amend commit message: %w", err)
	}

	return nil
}

func normalizeScope(scope, targetHash string) (string, string, error) {
	scope = strings.TrimSpace(scope)
	targetHash = strings.TrimSpace(targetHash)

	switch scope {
	case "single", "all":
		return scope, targetHash, nil
	case "specific":
		if targetHash == "" {
			return "", "", errors.New("target hash is required when scope is specific")
		}
		return scope, targetHash, nil
	default:
		return "", "", fmt.Errorf("invalid scope %q: expected single, specific, or all", scope)
	}
}

func buildTimeFilterScript(scope, targetHash, dateTime string) string {
	date := singleQuote(dateTime)
	setScript := fmt.Sprintf("export GIT_AUTHOR_DATE=%s; export GIT_COMMITTER_DATE=%s", date, date)
	if scope == "specific" {
		return fmt.Sprintf("case \"$GIT_COMMIT\" in %s*) %s ;; esac", targetHash, setScript)
	}
	return setScript
}

func buildAuthorFilterScript(scope, targetHash, name, email string) string {
	nameQ := singleQuote(name)
	emailQ := singleQuote(email)
	setScript := fmt.Sprintf(
		"export GIT_AUTHOR_NAME=%s; export GIT_AUTHOR_EMAIL=%s; export GIT_COMMITTER_NAME=%s; export GIT_COMMITTER_EMAIL=%s",
		nameQ, emailQ, nameQ, emailQ,
	)
	if scope == "specific" {
		return fmt.Sprintf("case \"$GIT_COMMIT\" in %s*) %s ;; esac", targetHash, setScript)
	}
	return setScript
}

func singleQuote(s string) string {
	return "'" + strings.ReplaceAll(s, "'", "'\"'\"'") + "'"
}

func (g *GitHistory) runCommandWithEnv(env map[string]string, args ...string) (stdout string, stderr string, err error) {
	cmd := exec.Command("git", args...)
	if g.Core.WorkDir != "" {
		cmd.Dir = g.Core.WorkDir
	}

	cmd.Env = append(cmd.Environ(), mapToEnvList(env)...)

	var outBuf bytes.Buffer
	var errBuf bytes.Buffer
	cmd.Stdout = &outBuf
	cmd.Stderr = &errBuf

	err = cmd.Run()
	return strings.TrimSpace(outBuf.String()), strings.TrimSpace(errBuf.String()), err
}

func mapToEnvList(env map[string]string) []string {
	if len(env) == 0 {
		return nil
	}

	out := make([]string, 0, len(env))
	for k, v := range env {
		out = append(out, fmt.Sprintf("%s=%s", k, v))
	}
	return out
}
